### ISSUE TITLE
After logout and login: Failed to initialize sandbox: getconf failed

### ISSUE THREAD

**Issue by kastiglione** [bazelbuild/bazel#7692]
### Description of the problem:

I ran a `bazel build` on macOS and got this error:

```
ERROR: Failed to initialize sandbox: getconf failed
```

I had just logged out and logged in, so the Bazel daemon existed, but any session state would have changed, if that matters. At first I had to kill with Activity Monitor, the bazel daemon wouldn't die when being killed with `kill -9`. After reading #2489 I found that `bazel shutdown` works.

This is reproducible for me.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

1. Run a `bazel build`
2. Log out and log in
3. Run `bazel build` again

### What operating system are you running Bazel on?

macOS 10.14.3

### What's the output of `bazel info release`?

release 0.23.1

###  Have you found anything relevant by searching the web?

Just #2489, but @philwo requested I create a new issue.

### Any other information, logs, or outputs that you want to share?

As mentioned above, `bazel shutdown` works around the problem.

**Comment by jmmv** [comment]
Interesting and, ugh, incomplete error messages that don't say *why* things failed. How often do you hit this though? Seems like quite a corner case.

**Comment by philwo** [comment]
I'm also quite surprised, that the Bazel server survives a logout - I always thought that logging out on macOS causes all processes launched during that user session to be terminated.

We should of course fix this and it's probably not too difficult.

**Comment by kastiglione** [comment]
I almost never logout, this is definitely an unlikely corner case. I created this issue just to document it, I personally don't need a fix.

**Comment by OleksiyA** [comment]
My OSX WindowsServer crashed couple times. I need to re-login, but all other processes were kept alive. Xcode would fail to build project afterwards.
My fix to run "bazel shutdown". After that all seems to work fine.

**Comment by evanj** [comment]
I agree this is somewhat unlikely, but I also just ran into this, using Bazel 3.2.0 on Mac OS X 10.15.6 FWIW, so this problem still exists. The `bazel shutdown` workaround is a good one. Possibly this Github issue is sufficient, since searching for the "Failed to initialize sandbox" error will lead the occasional user who hits it to find this workaround :)

**Comment by brentleyjones** [comment]
We've been encountering this as well, since WindowServer loves to crash on Catalina for us now.

**Comment by pcj** [comment]
Just hit this on macos catalina 10.15.7.

**Comment by pawaramol** [comment]
Just encountered this as well on MacOS Mojave 10.14.3 and bazel 3.7.1

**Comment by bigbangvn** [comment]
Encountered this on macos 10.15.7, bazel 3.1.0

**Comment by xhuoroku** [comment]
Add a data point, macOS 10.15.7, bazel 4.0.0-homebrew

**Comment by jerrymarino** [comment]
This one is still happening on Big Sur

**Comment by jschaf** [comment]
Present on Bazel 4.2.0 on macOS 11.2.2

**Comment by cmey** [comment]
Happened to me today on `bazel 4.2.1-homebrew` on macOS 11.6.1

**Comment by fernandoacorreia** [comment]
Happening to me on macOS 11.6.2 with bazel 4.2.2 (homebrew).

**Comment by keith** [comment]
Thanks for the reports folks, I think the next steps here are to:

1. Generally investigate if this happens with the `getconf` CLI as well
2. If it doesn't, investigate what the difference is between that CLI and the calls bazel does
3. If it does, investigate if there's a way around that mentioned in the manpage or anywhere else
4. If there's no workaround, investigate if there's another way to get the variables bazel gets from here, or if we can just replace those variables with something else.

If someone is motivated to take this on, that's where I would start!

**Comment by guillaumep** [comment]
> Happening to me on macOS 11.6.2 with bazel 4.2.2 (homebrew). The workaround was to reboot.

Happened to me also, I used the `bazel shutdown` command instead to resolve the issue.

**Comment by cassandrasandquist** [comment]
`macOS 12.4` and `bazel 5.2.0`

**Comment by zhuangya** [comment]
`macOS 13.0` and `bazel 5.0.0`

**Comment by dqhieu** [comment]
> > Happening to me on macOS 11.6.2 with bazel 4.2.2 (homebrew). The workaround was to reboot.
> 
> Happened to me also, I used the `bazel shutdown` command to resolve the issue.

`bazel shutdown` worked for me

**Comment by alexcstaley** [comment]
Encountered this today on `macOS 12.6.3` and `bazel 5.4.0`. Solved using `bazel shutdown` fix.

**Comment by BYVoid** [comment]
This bus is still reproducible by logging out and in.

**Comment by luispadron** [comment]
We've hit this a few times, definitely still any issue +1. Only a restart seems to fix in the cases we've hit.

**Comment by aaazzz** [comment]
I faced a similar issue. When I logged out and logged back in, the same error message appeared with `bazel build`.
Running `bazel shutdown` resolved the issue.

macos: `15.0.1 (24A348)`
bazel: `bazel 7.4.0-homebrew`

**Comment by Sacha924** [comment]
Got the same issue with the same error message : `ERROR: Failed to initialize sandbox: getconf failed`
`bazel shutdown` solves the problem
macos: `14.6.1`
bazel :  `6.5.0`

**Comment by mtullalizardi** [comment]
Got the same issue with the same error message as @Sacha924 . `bazel shutdown` fixed it.

**Comment by luispadron** [comment]
This is easily reproducible on macOS if you log out and log back in after having made a build. It seems the bazel server doesn't handle this well

**Comment by hachem-dd** [comment]
So.. this is still a problem I can assume

**Comment by akirchhoff-modular** [comment]
Ran into this on Bazel at commit 01705f6b0641140c93b2a12a88c69defbe82241c (9.0.0rc1-83-g01705f6b06)

**Comment by akirchhoff-modular** [comment]
> Thanks for the reports folks, I think the next steps here are to:
> 
> 1. Generally investigate if this happens with the `getconf` CLI as well

When this is occurring, the `getconf` CLI tool does not fail, at least for the examples given in the man page:

```
% getconf PATH    
/usr/bin:/bin:/usr/sbin:/sbin
% getconf NAME_MAX /tmp
255
```
  
> 2. If it doesn't, investigate what the difference is between that CLI and the calls bazel does
> 3. If it does, investigate if there's a way around that mentioned in the manpage or anywhere else
> 4. If there's no workaround, investigate if there's another way to get the variables bazel gets from here, or if we can just replace those variables with something else.
> 
> If someone is motivated to take this on, that's where I would start!

I do not have the bandwidth to test any of the rest of these.

**Comment by akirchhoff-modular** [comment]
I looked briefly in the Bazel source and the variables being accessed are:

* `DARWIN_USER_TEMP_DIR`
* `DARWIN_USER_CACHE_DIR`

The `getconf` CLI continues to work for these variables.

However, given that the problem is only when a Bazel server outlives a login session, it seems reasonable to conclude that Bazel is running these commands from that Bazel server after the session that created it has ended. I do not know, but I would guess that macOS deletes these two directories when a user is not logged in; or, to retrieve the value of these variables, it reaches out to a daemon that is no longer running for that user session.

So I guess then the question would be: What is the expected behavior of Bazel with regard to the temporary and cache directories it uses, when the temporary directory or cache directory has potentially changed? Four options seem plausible to me:

* Bazel queries these directories once it start up, and thereafter uses the cache values of these directories. It does not query them every single time it starts a sandbox.
* The Bazel client queries these directories on every invocation, and funnels those to the Bazel server, where they are used during sandbox creation. The Bazel server never directly runs getconf.
* The Bazel Darwin sandbox is changed such that it no longer carves out exceptions for these two paths when creating a sandbox, and hence the two paths are not needed.
* Bazel servers are tied to a user session and shut down during log out.

I don't know how to implement any of these, however.

**Comment by keith** [comment]
https://github.com/bazelbuild/bazel/pull/29131

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 9.1.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/9.1.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=9.1.0rc1`. Thanks!

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 8.7.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/8.7.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=8.7.0rc1`. Thanks!
